### PR TITLE
Sm/update

### PR DESCRIPTION
**Pull request by NelsonVides**
This PR addresses changes in the [Stream Management XEP](https://xmpp.org/extensions/xep-0198.html).

Proposed changes are (hopefully) very well explained in the tests:

* `client_enables_sm_twice_fails_with_correct_error_stanza`: the new XEP _requires_ enabling be requested only once. If a client requests so twice, the client should be considered faulty and closed gracefully.
* `session_resumed_then_old_session_is_closed_gracefully_with_correct_error_stanza`: We were not testing what happened to the old session on resumption apart from being closed. The XEP requires that if the connection is still there, a `<conflict>` stream error be send and connection be closed gracefully.
* `session_resumed_and_old_session_dead_doesnt_route_error_to_new_session`: verifies that if the old session cannot receive the `<conflict>` stanzas, they're not rerouted to the newly connected session anyway.
* `h_non_given_closes_stream_gracefully` is a test that should have been there before. The XEP _requires_ the `h` attribute to be given. When it's not, the connection is faulty and should be closed. We were doing too happily a `binary_to_integer(xml:get_tag_attr_s(<<"h">>, El))` without catching the possible error, which was then provoking the `c2s` process to die ungracefully. Now this error is caught and a proper stream:error is send together with a stream closing.
* One more old test had to change, `client_acks_more_than_sent`, as the expected error that this test checks was checking is also now differently defined.

A todo about timestamps on rerouting was removed, because they're already added in https://github.com/esl/MongooseIM/blob/b288406a5f0ea70b04fbc3e218f4436bbd79cbd3/src/ejabberd_c2s.erl#L1147, when the execution path reaches `buffer_out_stanza`, so no need to worry if it was done or not, it was always done indeed.

PS: There's one more feature in SM coming in, but it's all complex by itself so it will come in a different PR.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7214.1](https://travis-ci.org/esl/MongooseIM/jobs/587506295) / Erlang 22.0 / small_tests / b942adffd3a4d1ba03ff058939ba7b01da77acc7
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2468/7214/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2468/7214/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-20_14.20.55/lib.mongooseim.logs/run.2019-09-20_14.20.55/suite.log.html)

File: src/ejabberd_c2s.erl [hunk 1]
```diff
@@ -2800,17 +2805,26 @@ stream_mgmt_handle_ack(NextState, El, #state{} = SD) ->
             maybe_send_element_from_server_jid_safe(SD, mongoose_xmpp_errors:invalid_namespace()),
             maybe_send_trailer_safe(SD),
             {stop, normal, SD};
-        throw:{policy_violation, Reason} ->
-            PolicyViolation = mongoose_xmpp_errors:policy_violation(SD#state.lang, Reason),
-            maybe_send_element_from_server_jid_safe(SD, PolicyViolation),
+        error:badarg ->
```
Comment (arcusfelis) [inline src/ejabberd_c2s.erl]: isn't it too wide matching?
Maybe introducing parse_h, that throws some error we can match would be nicer.

**Comment by mongoose-im** [comment]
[7216.1](https://travis-ci.org/esl/MongooseIM/jobs/587529439) / Erlang 22.0 / small_tests / 165ba0dd2af4e21ec09ad87f04b4bda512d03639
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2468/7216/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2468/7216/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-20_16.36.28/lib.mongooseim.logs/run.2019-09-20_16.36.28/suite.log.html)

**Comment by mongoose-im** [comment]
[7223.1](https://travis-ci.org/esl/MongooseIM/jobs/587783745) / Erlang 22.0 / small_tests / eb11cff68020d28212fc55c3bbeccb4228163bd6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2468/7223/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2468/7223/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-21_08.26.08/lib.mongooseim.logs/run.2019-09-21_08.26.09/suite.log.html)

File: big_tests/tests/sm_SUITE.erl [hunk 2]
```diff
@@ -249,13 +295,37 @@ h_ok_after_a_chat(ConfigIn) ->
         escalus:send(Alice, escalus_stanza:sm_ack(3 + NDiscarded))
     end).
 
+h_non_given_closes_stream_gracefully(ConfigIn) ->
+    AStanza = #xmlel{name = <<"a">>,
+               attrs = [{<<"xmlns">>, <<"urn:xmpp:sm:3">>}]},
+    Config = escalus_users:update_userspec(ConfigIn, alice,
+                                           stream_management, true),
+    escalus:fresh_story(Config, [{alice,1}], fun(Alice) ->
+        %% Ack, so that unacked messages don't go into offline store.
```
Comment (fenek) [inline big_tests/tests/sm_SUITE.erl]: Copy & paste? :)

File: big_tests/tests/sm_SUITE.erl [hunk 3]
```diff
@@ -985,6 +1067,10 @@ wait_until_disconnected(UserSpec) ->
     mongoose_helper:wait_until(fun() -> get_user_alive_resources(UserSpec) =:= [] end, true,
                                #{name => get_user_alive_resources}).
 
+wait_until_client_disconnects(Client) ->
```
Comment (fenek) [inline big_tests/tests/sm_SUITE.erl]: https://github.com/esl/escalus/blob/master/src/escalus_connection.erl#L418

File: src/ejabberd_c2s.erl [hunk 4]
```diff
@@ -3104,8 +3124,7 @@ maybe_add_timestamp({F, T, #xmlel{name= <<"message">>}=Packet}=PacketTuple, Time
         <<"headline">> ->
             PacketTuple;
         _ ->
-            %% TODO: ?MYNAME (or server taken from c2s state) not <<"localhost">>
```
Comment (fenek) [inline src/ejabberd_c2s.erl]: TODO should remain here, as actually the right domain to use is the server from c2s state.

File: big_tests/tests/sm_SUITE.erl [hunk 5]
```diff
@@ -177,6 +181,48 @@ server_returns_failed(Config, ConnActions) ->
...
+    Steps2 = connection_steps_to_stream_resumption(SMID, SMH),
+    {ok, Alice2, _} = escalus_connection:start(AliceSpec, Steps2),
+    process_initial_stanza(Alice2),
+    %% THEN: Old session is gracefully closed with the correct error stanza
+    escalus:assert(is_stream_error, [<<"conflict">>, <<>>],
+                   escalus_connection:get_stanza(Alice, close_old_stream)),
+    escalus:assert(is_stream_end,
+                   escalus_connection:get_stanza(Alice, close_old_stream)),
+    wait_until_client_disconnects(Alice),
+    true = escalus_connection:is_connected(Alice2),
```
Comment (arcusfelis) [inline big_tests/tests/sm_SUITE.erl]: The best way to check if connection is still working is to send and receive stanza between two users.

**Comment by mongoose-im** [comment]
[7224.1](https://travis-ci.org/esl/MongooseIM/jobs/588386682) / Erlang 22.0 / small_tests / 021523f6741b013ab0ede39adeba84706fe44757
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2468/7224/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2468/7224/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-23_10.14.53/lib.mongooseim.logs/run.2019-09-23_10.14.53/suite.log.html)

File: src/ejabberd_c2s.erl (hunk 4 above)
Comment (arcusfelis) [reply src/ejabberd_c2s.erl]: just use SD#state.server here (pass Server as an extra argument)

**Pull request merged**